### ISSUE TITLE
Rate Limiting?

### ISSUE THREAD

**Issue by Sm0k3scr33n** [yyyar/gobetween#266]
Hello Maintainers,

I was wondering if rate limiting or similar feature is available or if there are plans to make it available?

**Comment by nickdoikov** [comment]
@Sm0k3scr33n 
please clarify what exactly you mean ?
we currently have a connection limit per server but ,as far as i understand, you are talking about rate limiting per client ip  , correct?

**Comment by nickdoikov** [comment]
So i think it needs to be implemented this way to cover all needs
1) connection limit - limit connections(TCP sessions) to N from each ip that initiating a connection to a Server listener .
It will works like a counter . 
N connection per second or minute 
2) Rate Limit - limit amount of incoming packets(TCP/UDP) from each ip that initiating a connection to a Server listener .

If connection/rate limit reached for M time (minutes/seconds) exec some script this way:
limit_action_script = ["/path/to/script", "arg1", "arg2"] 
GB should add IP that rate limited as a first script argument

after connection rate limit decreased and not increased during T timeframe (in seconds/minutes) exec different script this way: 
delimit_action_script = ["/path/to/script", "arg1", "arg2"] 
GB should add IP that rate limited as a first script argument

such actions needed for perform actions like add rule to external fw (Cloud provider  security groups for example), or execute some another logic for attack mitigation

@illarion @yyyar  what do you think about this logic?

**Comment by Sm0k3scr33n** [comment]
> @Sm0k3scr33n
> please clarify what exactly you mean ?
> we currently have a connection limit per server but ,as far as i understand, you are talking about rate limiting per client ip , correct?

Yes sir, as you outlined, rate limit per client IP in a given period of time would be the use case I am most interested in. While we would be able to handle lots of traffic from many IP's, we would place a rate limit to help mitigate over use and to force compliance in our own ULA for client api usage. Thank you very much :) Cheers

**Comment by Sm0k3scr33n** [comment]
Jotting out more thoughts on this
![image](https://user-images.githubusercontent.com/7565846/74963619-43db4280-53cf-11ea-81a0-de939965e4e8.png)

I would like to avoid using a database, but mongo is fast:
![image](https://user-images.githubusercontent.com/7565846/74963650-535a8b80-53cf-11ea-91da-e44eb5bd8f9f.png)
